examples/nn.dx

```diff
@@ -16,39 +16,39 @@ instance [Add a, Add b] Add (a & b)
 instance [VSpace a, VSpace b] VSpace (a & b)
   scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
 
-def Layer (inp:Type) (out:Type) (param:Type) : Type=
-  {forward: param -> inp -> out &
-   init: Key -> param
-  }
+data Layer inp:Type out:Type params:Type =
+  AsLayer {forward:(params -> inp -> out) & init:(Key -> params)}
```

Comment:
How about `MkLayer` instead of `Layer` as the constructor name?

I would also be happy with changing the _constructor name_ `AsList` to `MkList`. @dougalm

Reply:
I don't think we have to always stick to a single convention. `AsLayer` and `AsList` both sound good to me.

Reply:
On second thought: totally fair.

My original objection is that `As` sounds like type-casting (e.g. `@` in `0@(Fin 10)`), which is different.
But `AsLayer` and `AsList` really just perform type erasure and bundling, which is not so far from casting.